src/autopas/selectors/tuningStrategy/PredictiveTuning.h

```diff
@@ -183,32 +245,227 @@ void PredictiveTuning::selectOptimalSearchSpace() {
+      predictionOutput(configuration);
+    } else if (_traversalTimesStorage[configuration].size() >= _testsUntilFirstPrediction) {
+      const auto &vector = _traversalTimesStorage[configuration];
+      const auto &[traversal1Iteration, traversal1Time] = vector[vector.size() - 1];
+      const auto &[traversal2Iteration, traversal2Time] = vector[vector.size() - 2];
+
+      const auto gradient = (traversal1Time - traversal2Time) / (traversal1Iteration - traversal2Iteration);
+      const auto delta = _iterationBeginTuningPhase - traversal1Iteration;
+
+      // time1 + (time1 - time2) / (iteration1 - iteration2) / tuningPhase - iteration1)
+      _configurationPredictions[configuration] = traversal1Time + gradient * delta;
+      _configurationPredictionFunction[configuration].clear();
+      _configurationPredictionFunction[configuration].emplace_back(gradient);
+      _configurationPredictionFunction[configuration].emplace_back(traversal1Time);
+      predictionOutput(configuration);
+    } else {
+      // When a configuration was not yet tested twice.
+      _configurationPredictions[configuration] = std::numeric_limits<long unsigned int>::max();
+      _tooLongNotTestedSearchSpace.emplace(configuration);
+      AutoPasLog(debug, "No traversal time prediction for {}", configuration.toString());
+    }
+  }
+}
+
+void PredictiveTuning::linearRegression() {
+  for (const auto &configuration : _searchSpace) {
+    // if configuration was not tested in last tuning phase reuse prediction function.
+    if ((_lastTest[configuration] != (_tuningIterationsCounter - 1)) and
+        _configurationPredictionFunction[configuration].size() == 2) {
+      // gradient * iteration + y-intercept
+      _configurationPredictions[configuration] =
+          _configurationPredictionFunction[configuration][0] * _iterationBeginTuningPhase +
+          _configurationPredictionFunction[configuration][1];
+      predictionOutput(configuration);
+    } else if (_traversalTimesStorage[configuration].size() >= _testsUntilFirstPrediction) {
+      // TODO: Maybe look at casting again - already better than before
+      size_t iterationMultTime = 0, iterationSum = 0, iterationSquareSum = 0, timeSum = 0;
+
+      for (auto i = _traversalTimesStorage[configuration].size() - _testsUntilFirstPrediction;
+           i < _traversalTimesStorage[configuration].size(); i++) {
+        const auto &[iteration, time] = _traversalTimesStorage[configuration][i];
+        iterationMultTime += iteration * time;
+        iterationSum += iteration;
+        iterationSquareSum += iteration * iteration;
+        timeSum += time;
+      }
+
+      const double iterationMeanValue =
+          static_cast<double>(iterationSum) / static_cast<double>(_testsUntilFirstPrediction);
+      const auto timeMeanValue = timeSum / _testsUntilFirstPrediction;
+
+      const auto numerator = static_cast<long>(iterationMultTime) - static_cast<long>(iterationSum * timeMeanValue);
```

Comment:
```suggestion
      // cast unsigned to signed because this difference can be negative
      const auto numerator = static_cast<long>(iterationMultTime) - static_cast<long>(iterationSum * timeMeanValue);
```
add comments why casts are necessary